### ISSUE TITLE
Workflow Deployments Use Image Tag :latest

### ISSUE THREAD

**Issue by patgmiller** [argoproj/argo#2715]
Checklist:

 * [x] I've included the version. v2.6.3
 * [x] I've included reproduction steps.
 * [ ] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**: I deployed version v2.6.3 to our company's development k8s cluster and a few weeks later the workflows stopped working with error `Failed to list *v1alpha1.ClusterWorkflowTemplate: clusterworkflowtemplates.argoproj.io is forbidden: User`

**What you expected to happen**: I expect not to have newer versions of the product show up unless I specifically apply them.

**How to reproduce it (as minimally and precisely as possible)**: Deploy a previous version of argo workflows and submit a workflow. The deployments will have the `:latest` image version.

**Anything else we need to know?**:

**Environment**:
- Argo Workflow Version: https://github.com/argoproj/argo/manifests/cluster-install?ref=v2.6.3

- Argo version: v2.7.0
```
$ argo version
argo: v2.7.0-rc2
  BuildDate: 2020-03-23T17:54:57Z
  GitCommit: 240d7ad9298c60a69d4ce056e3d83ef9283a83ec
  GitTreeState: clean
  GitTag: v2.7.0-rc2
  GoVersion: go1.13.4
  Compiler: gc
  Platform: darwin/amd64
```
- Kubernetes version : v1.14.10-gke.27
```
$ kubectl version -o yaml
clientVersion:
  buildDate: "2019-10-15T23:41:55Z"
  compiler: gc
  gitCommit: c97fe5036ef3df2967d086711e6c0c405941e14b
  gitTreeState: clean
  gitVersion: v1.16.2
  goVersion: go1.12.10
  major: "1"
  minor: "16"
  platform: darwin/amd64
serverVersion:
  buildDate: "2020-02-21T18:01:40Z"
  compiler: gc
  gitCommit: 145f9e21a4515947d6fb10819e5a336aff1b6959
  gitTreeState: clean
  gitVersion: v1.14.10-gke.27
  goVersion: go1.12.12b4
  major: "1"
  minor: 14+
  platform: linux/amd64
```

**Other debugging information (if applicable)**:

- workflow-controller logs:
```
alsq@UL1804VM:~/bin$ kubectl -n argo logs -f workflow-controller-74b4f4659b-w2d8c
[ excerpt ]
E0413 18:20:41.979797       1 reflector.go:125] pkg/mod/k8s.io/client-go@v0.0.0-20191225075139-73fd2ddc9180/tools/cache/reflector.go:98: Failed to list *v1alpha1.ClusterWorkflowTemplate: clusterworkflowtemplates.argoproj.io is forbidden: User "system:serviceaccount:argo:argo" cannot list resource "clusterworkflowtemplates" in API group "argoproj.io" at the cluster scope
[ repeating ]
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by alexec** [comment]
This should not happen obviously. Can I ask questions please?

The executor image can be configure either in the config map or the deployment. Can you please attach:

```
kubectl get deploy workflow-controller
kubectl get cm workflow-controller-configmap
```

You should not see latest there.

Can you also attach the URL to the YAMLs you deployed? It should be https://raw.githubusercontent.com/argoproj/argo/v2.6.3/manifests/install.yaml but I cannot see latest there.

**Comment by simster7** [comment]
Just adding some potentially useful information in debugging this.

Another user had installed Argo from the `stable` tag, which had been pointing to a random commit between releases (hence `:latest` being on the manifests): https://github.com/argoproj/argo/issues/2693#issuecomment-614182828

Perhaps something similar happened here?

**Comment by patgmiller** [comment]
Here are those requested k8s objects. 

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "4"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"name":"workflow-controller","namespace":"argo"},"spec":{"selector":{"matchLabels":{"app":"workflow-controller"}},"template":{"metadata":{"labels":{"app":"workflow-controller"}},"spec":{"containers":[{"args":["--configmap","workflow-controller-configmap","--executor-image","argoproj/argoexec:v2.6.3"],"command":["workflow-controller"],"image":"argoproj/workflow-controller:v2.6.3","name":"workflow-controller"}],"serviceAccountName":"argo"}}}}
  creationTimestamp: "2020-03-30T15:03:20Z"
  generation: 5
  labels:
    app: workflow-controller
  name: workflow-controller
  namespace: argo
  resourceVersion: "385162250"
  selfLink: /apis/extensions/v1beta1/namespaces/argo/deployments/workflow-controller
  uid: 97e5e0d0-7297-11ea-8a50-42010a8001fc
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: workflow-controller
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: workflow-controller
    spec:
      containers:
      - args:
        - --configmap
        - workflow-controller-configmap
        - --executor-image
        - argoproj/argoexec:v2.6.3
        command:
        - workflow-controller
        image: argoproj/workflow-controller:v2.6.3
        imagePullPolicy: IfNotPresent
        name: workflow-controller
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: argo
      serviceAccountName: argo
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2020-04-13T20:10:11Z"
    lastUpdateTime: "2020-04-13T20:10:11Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2020-03-30T15:03:20Z"
    lastUpdateTime: "2020-04-13T20:33:08Z"
    message: ReplicaSet "workflow-controller-848566d646" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 5
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```

```yaml
apiVersion: v1
data:
  config: |-
    artifactRepository:
      s3:
        accessKeySecret:
          name: argo-artifact-creds
          key: accessKey
        secretKeySecret:
          name: argo-artifact-creds
          key: secretKey
        endpoint: storage.googleapis.com
        bucket: argo-artifacts-dev-1
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"config":"artifactRepository:\n  s3:\n    accessKeySecret:\n      name: argo-artifact-creds\n      key: accessKey\n    secretKeySecret:\n      name: argo-artifact-creds\n      key: secretKey\n    endpoint: storage.googleapis.com\n    bucket: argo-artifacts-dev-1"},"kind":"ConfigMap","metadata":{"annotations":{},"name":"workflow-controller-configmap","namespace":"argo"}}
  creationTimestamp: "2020-03-30T15:03:18Z"
  name: workflow-controller-configmap
  namespace: argo
  resourceVersion: "374737816"
  selfLink: /api/v1/namespaces/argo/configmaps/workflow-controller-configmap
  uid: 96e2b5c5-7297-11ea-8a50-42010a8001fc
```

And some additional info, I am using `kustomize` to deploy this. Here is my `kustomization.yaml` file.

```yaml
# base/kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namespace: argo

resources:
- resources/argo-artifact-creds.secret.yaml
- resources/argo-workflow-serviceaccount.yaml
- resources/argo-workflow-role.yaml
- resources/argo-workflow-role-binding.yaml
- github.com/argoproj/argo/manifests/cluster-install?ref=v2.6.3
- github.com/argoproj/argo-events/hack/k8s/manifests?ref=v0.13.0

patchesStrategicMerge:
  - overlays/argo-events-serviceaccount.yaml
  - overlays/gateway-controller-configmap.yaml
  - overlays/sensor-controller-configmap.yaml
  - overlays/argo-server-deployment.yaml
  - overlays/workflow-controller-deployment.yaml
```

After I ran into this issue I applied the following diff, originally every image for argo workflows was using `:latest` which can be seen at the specific tagged version of `v2.6.3`

[argo-wf-set-image-tags-v2.6.3.txt](https://github.com/argoproj/argo/files/4488826/argo-wf-set-image-tags-v2.6.3.txt)

**Comment by alexec** [comment]
I think this is what's happening. Rather that using the manifests listed [here](https://github.com/argoproj/argo/tree/master/manifests) you're directly referencing the source code because you are using Kustomize. 

Since v2.5 these `manifests/cluster-install` and `manifests/install` contain `latest`, not the release tag.

It's open to discussion as to whether we support this. 

# Pro

* Kustomize is a commonly used config management tool we might wish to formally support.
* Regardless of whether or not people should use this, they will.

If we decide to support this, we should:

* Fix the manifests and back-port.
* Put in place guard rails.
* Add these files to the public API.
* Document it as a supported installation option.

# Contra

* We do not document this as an installation option, nor do we mention it in the release notes.
* This is a risky method - not suitable for production - because tags can move and without taking steps to reduce this risk.
* As this is the first time someone has raised this as an issue, we could conclude it is not commonly used.

If we do not support it, then we should:

* Add a warning not to use this method.
* Add the correct manifests to the public API.

**Comment by alexec** [comment]
@alexmt @jessesuen can I get your thought on whether or not we support Kustomize as an installation option.

**Comment by ptemmer** [comment]
We are also using Kustomize to deploy Argo. What we do is overwrite the image in our overlay:
```
images:
- name: argoproj/workflow-controller
  newTag: v2.7.2
```
However, I must admit that when pulling in a resource from a specific github tag, one would expect the containers included in the deployments to also match that tag version.

```
bases:
- git@github.com:argoproj/argo/manifests/namespace-install/?ref=v2.7.2
```

**Comment by alexec** [comment]
We will fix this and back-port to v2.7.5

**Comment by patgmiller** [comment]
Thanks I am new to managing k8s objects with `kustomize`, thanks for helping me understand the issue. And thank you for adjusting the workflow to support it.
